### PR TITLE
Tracing tutorial

### PR DESCRIPTION
**Pull request by mturilli**
Adds the tutorial about tracing with RP and the event model, addressing [#2532](https://github.com/radical-cybertools/radical.pilot/issues/2532)

### CODE REVIEW

File: docs/source/internals.rst [hunk 1]
```diff
@@ -101,83 +83,326 @@ Task
...
+   ve_update_start     : pilot ve update      starts                (uid: pilot, [CFG-R])
+   ve_update_start     : pilot ve update      stops                 (uid: pilot, [CFG-R])
+   ve_create_stop      : pilot ve creation    stops                 (uid: pilot, [CFG-R])
+   rp_install_start    : rp stack install     starts                (uid: pilot, [CFG-R])
+   rp_install_stop     : rp stack install     stops                 (uid: pilot, [CFG-R])
+   ve_setup_stop       : pilot ve setup       stops                 (uid: pilot, [CFG-R])
+   ve_activate_start   : pilot ve activation  starts                (uid: pilot, [CFG-R])
+   ve_activate_start   : pilot ve activation  stops                 (uid: pilot)
+   client_barrier_start: wait for client signal                     (uid: pilot, [CFG-R])
+   client_barrier_stop : client signal received                     (uid: pilot, [CFG-R])
```
Comment (mtitov) [inline docs/source/internals.rst:231]: `client_barrier_*` obsolete events (to be removed)

File: docs/source/internals.rst [hunk 2]
```diff
@@ -101,83 +83,326 @@ Task
...
+   sync_rel            : time sync event                            (uid: pilot, msg: 'agent_0 start')
```
Comment (mtitov) [inline docs/source/internals.rst:232]: obsolete event

File: docs/source/internals.rst [hunk 3]
```diff
@@ -101,83 +83,326 @@ Task
...
+   staging_out_stop    : staging request stops                      (uid: task, msg: did, [Task-DS])
+
+   partial orders
+   * per task          : staging_stdout_start, staging_stdout_stop,
+                         staging_stderr_start, staging_stderr_stop,
+                         staging_uprof_start,  staging_uprof_stop,
+   * per file          : staging_out_skip \
+                       | (staging_out_start, (staging_out_fail | staging_out_stop))
+
+UMGRStagingOutputComponent (Component)
```
Comment (mtitov) [inline docs/source/internals.rst:357]: `TMGRStagingOutputComponent`

File: docs/source/internals.rst [hunk 4]
```diff
@@ -101,83 +83,326 @@ Task
...
+   partial orders
+   * per state update  : update_request, update_pushed
+
+All profiles
+------------
+
+::
+
+   sync_abs            : sets an absolute, NTP synced time stamp               ([INTERNAL])
+   sync_rel            : sets a *pair* of time stamps considered simultaneous  ([INTERNAL])
```
Comment (mtitov) [inline docs/source/internals.rst:386]: `sync_rel` looks like it is obsolete everywhere (some other components above also referred it) and we only use `sync_abs`

**Review by mtitov — COMMENTED** [review]
Comments regarding events

**Review by mtitov — COMMENTED** [review]
couple comments regarding notebook

File: docs/source/internals.rst (hunk 2 above)
Comment (mturilli) [reply docs/source/internals.rst]: I assumed that referred to `sync_rel`

File: docs/source/internals.rst (hunk 4 above)
Comment (mturilli) [reply docs/source/internals.rst]: I removed all the `sync_rel`

File: docs/source/tutorials/profiling.ipynb [hunk 5]
```diff
@@ -5,11 +5,375 @@
...
+    "- `state` = (facultative) the name given to the event when it is also a state, i.e., guaranteed to be triggered in every configuration of RP\n",
+    "- `msg` = (facultative) an arbitrary message associated to that event.\n",
+    "\n",
+    "For example, here is the record of the `launch_start` event for a task:\n",
+    "\n",
+    "```\n",
+    "1684787127.2228520,launch_start,,MainThread,task.000000,AGENT_EXECUTING,\n",
+    "```\n",
+    "\n",
+    "As you will see later in this tutorial, you can use `launch_start` to monitor when a task is/was executed. Here is another example, this time for the `schedule_ok` event triggered by RP's Agent Scheduler:\n",
```
Comment (mtitov) [inline docs/source/tutorials/profiling.ipynb:60]: `to monitor when a task is/was executed` -> `to monitor when a task is/was launched to be executed`?

File: docs/source/tutorials/profiling.ipynb [hunk 6]
```diff
@@ -5,11 +5,375 @@
...
+    "\n",
+    "</div>\n",
+    "\n",
+    "\n",
+    "## RADICAL Analytics\n",
+    "\n",
+    "Parsing and filtering the `*.prof`files requires a solid understanding of: \n",
+    "\n",
+    "1. RP's event/state model in order to select only the information you actually need;\n",
+    "2. the format in which the information is encoded within in `*.prof` file;\n",
```
Comment (mtitov) [inline docs/source/tutorials/profiling.ipynb:80]: ``encoded within in `*.prof` file`` -> ``encoded within `*.prof` file`` (remove `in`)?

File: docs/source/tutorials/profiling.ipynb [hunk 7]
```diff
@@ -5,11 +5,375 @@
...
+    "That is both **difficult** and **cumbersome**. \n",
+    "\n",
+    "In order to facilitate using RP traces to profile the execution of your application, we created [RADICAL Analytics (RA)](https://radicalanalytics.readthedocs.io/en/latest/). Please see the linked documentation and consider using RA whenever you need postmortem profiling. In the following, we discuss how you can perform some task-related tracing at runtime or postmortem, i.e., while RP is executing your application or once it finished. This is useful while developing your application and/or for sport-checks while executing a large run. For production profiling, you should use RA. \n",
+    "\n",
+    "## Tracing at runtime\n",
+    "\n",
+    "In a real-life scenario, you will:\n",
+    "\n",
+    "1. Start the execution of your application on a [supported](../supported.rst) HPC platform;\n",
+    "2. while the application is running, execute shell commands to parse and filter across multiple `*.profile` files.\n",
```
Comment (mtitov) [inline docs/source/tutorials/profiling.ipynb:93]: `` `*.profile` files`` -> `` `*.prof` files``?

File: docs/source/tutorials/profiling.ipynb [hunk 8]
```diff
@@ -5,11 +5,375 @@
...
+    "Parsing and filtering the `*.prof`files requires a solid understanding of: \n",
+    "\n",
+    "1. RP's event/state model in order to select only the information you actually need;\n",
+    "2. the format in which the information is encoded within in `*.prof` file;\n",
+    "3. in which file or files RP recorded the information you need;\n",
+    "4. bash commands/scripts to parse and filter those files.\n",
+    "\n",
+    "That is both **difficult** and **cumbersome**. \n",
+    "\n",
+    "In order to facilitate using RP traces to profile the execution of your application, we created [RADICAL Analytics (RA)](https://radicalanalytics.readthedocs.io/en/latest/). Please see the linked documentation and consider using RA whenever you need postmortem profiling. In the following, we discuss how you can perform some task-related tracing at runtime or postmortem, i.e., while RP is executing your application or once it finished. This is useful while developing your application and/or for sport-checks while executing a large run. For production profiling, you should use RA. \n",
```
Comment (mtitov) [inline docs/source/tutorials/profiling.ipynb:86]: `for sport-checks` -> `for spot-checks`?

File: docs/source/tutorials/profiling.ipynb [hunk 9]
```diff
@@ -5,11 +5,375 @@
...
+    "</div>\n",
+    "\n",
+    "Here we: \n",
+    "\n",
+    "1. Use the `RADICAL_PROFILE` environment variable to enable RP tracing;\n",
+    "2. run a condensed version of the application we used for the [Getting Started](../getting_started.ipynb).\n",
+    "\n",
+    "<div class=\"alert alert-warning\">\n",
+    "    \n",
+    "__Warning:__ If you do not set `RADICAL_PROFILE`, RP will **not** create the `*.prof` files.\n",
```
Comment (mtitov) [inline docs/source/tutorials/profiling.ipynb:110]: by default `RADICAL_PROFILE` is set to `TRUE`

https://github.com/radical-cybertools/radical.utils/blob/devel/src/radical/utils/configs/utils_default.json#L9

**Review by mtitov — COMMENTED** [review]
more comments (final ones)

File: docs/source/tutorials/profiling.ipynb [hunk 10]
```diff
@@ -5,11 +5,375 @@
...
+    "\n",
+    "Where:\n",
+    "\n",
+    "- `time` = (mandatory) timestamp expressed in Unix Time (epoch)\n",
+    "- `event` = (mandatory) the name of the timestamped event\n",
+    "- `comp` = (facultative) the name of the component that triggered the event\n",
+    "- `thread` = (mandatory) the ID of the thread in which the event was triggered with `comp`\n",
+    "- `uid` = (mandatory) the ID of the entity to which that event belongs (e.g., task or pilot)\n",
+    "- `state` = (facultative) the name given to the event when it is also a state, i.e., guaranteed to be triggered in every configuration of RP\n",
+    "- `msg` = (facultative) an arbitrary message associated to that event.\n",
```
[the remark this answers is not in the document]
Comment (mturilli) [reply docs/source/tutorials/profiling.ipynb:52]: Went for `optional`

File: docs/source/tutorials/profiling.ipynb (hunk 9 above)
Comment (mtitov) [reply docs/source/tutorials/profiling.ipynb]: Confirm about default value - I didn't set it and prof-files were created, and agent side had a corresponding env variable set
```
> cat ~/radical.pilot.sandbox/re.session.polaris-login-04.matitov.019515.0009/pilot.0000/env/bs0_pre_0.env | grep RADICAL
RADICAL_BASE=/home/matitov/radical.pilot.sandbox
RADICAL_PROFILE=TRUE
RADICAL_SMT=1
```

**Review by mtitov — APPROVED** [review]
LGTM!

**Pull request merged**